### PR TITLE
Update to wabt 1.0.6

### PR DESCRIPTION
**Pull request by axic**
1.0.6 gives us two benefits (amongst others):
- https://github.com/WebAssembly/wabt/pull/886: fixes host trapping
- https://github.com/WebAssembly/wabt/pull/882: speeds up the build process

### CODE REVIEW

**Comment by axic** [comment]
Feel free, not sure why it doesn't. Had a look at the wabt commits and there wasn't any change in the cmake for this, though I haven't done an actual diff on the release yet.

**Comment by chfast** [comment]
I fixed the wabt build, but includes are wrong now.

**Comment by axic** [comment]
Perhaps we need to fix upstream? :(

**Comment by chfast** [comment]
> Perhaps we need to fix upstream? :(

Nothing to fix upstream. Just the API has changed a lot.

**Comment by axic** [comment]
I'd like to split this up before it's merged.

**Comment by axic** [comment]
I'm not a fan of the reformat step.

**Comment by chfast** [comment]
> I'm not a fan of the reformat step.

Removed.

File: src/wabt.cpp [hunk 1]
```diff
@@ -17,12 +17,12 @@
 #include <vector>
 #include <iostream>
 
-#include "src/binary-reader-interp.h"
+#include "src/interp/binary-reader-interp.h"
 #include "src/binary-reader.h"
 #include "src/cast.h"
-#include "src/error-handler.h"
+//#include "src/error-handler.h"
```
Comment (axic) [inline src/wabt.cpp]: Stray include.

**Comment by axic** [comment]
Rebased to trigger mac build.

**Comment by chfast** [comment]
> Rebased to trigger mac build.

You need to pass `-DHERA_WABT=ON` to macos build.

File: src/wabt.cpp [hunk 2]
```diff
@@ -188,329 +71,6 @@ unique_ptr<WasmEngine> WabtEngine::create()
...
-  const ErrorCallback& callback
-) {
-  (void)import;
-  (void)func;
-  (void)callback;
-  HERA_DEBUG << "Importing " << import->field_name << "\n";
-  interp::HostFunc *hostFunc = reinterpret_cast<interp::HostFunc*>(func);
-  if (import->field_name == "useGas") {
-    if (func_sig->param_types.size() != 1 || func_sig->result_types.size() != 0)
-      return Result::Error;
```
Comment (axic) [inline src/wabt.cpp]: All of these arg/result checks are removed. Should be kept.

**Comment by axic** [comment]
Rebased this and should be ok to merge [once this comment is addressed](https://github.com/ewasm/hera/pull/454#discussion_r241500468).

File: src/wabt.cpp [hunk 3]
```diff
@@ -520,21 +80,151 @@ ExecutionResult WabtEngine::execute(
...
-  // Lets add our host module
+  // Create host module
   // The lifecycle of this pointer is handled by `env`.
   interp::HostModule* hostModule = env.AppendHostModule("ethereum");
   heraAssert(hostModule, "Failed to create host module.");
-  hostModule->import_delegate = unique_ptr<WabtEthereumInterface>(interface);
 
+  hostModule->AppendFuncExport(
+    "useGas",
+    {{Type::I64}, {}},
```
Comment (chfast) [inline src/wabt.cpp]: WABT handles the argument checking now. Here it is declared that `useGas()` takes single `i64` arg and returns nothing.

File: src/wabt.cpp (hunk 2 above)
Comment (axic) [reply src/wabt.cpp]: Solved: https://github.com/ewasm/hera/pull/454#discussion_r241505864

File: src/wabt.cpp [hunk 4]
```diff
@@ -520,21 +80,151 @@ ExecutionResult WabtEngine::execute(
...
+  // Set up the wabt Environment, which includes the Wasm store
+  // and the list of modules used for importing/exporting between modules
   interp::Environment env;
 
-  // Lets instantiate our state
+  // Set up interface to eei host functions
   ExecutionResult result;
-
   // FIXME: shouldn't have this loose pointer here, but needed for setWasmMemory
   WabtEthereumInterface* interface = new WabtEthereumInterface{context, state_code, msg, result, meterInterfaceGas};
```
Comment (axic) [inline src/wabt.cpp]: I thought I have left this comment here a week ago: this is leaking memory now.

@chfast this was the reason I didn't merge it.

File: src/wabt.cpp [hunk 5]
```diff
@@ -520,21 +80,151 @@ ExecutionResult WabtEngine::execute(
...
 
-  // Lets add our host module
+  // Create host module
   // The lifecycle of this pointer is handled by `env`.
   interp::HostModule* hostModule = env.AppendHostModule("ethereum");
   heraAssert(hostModule, "Failed to create host module.");
-  hostModule->import_delegate = unique_ptr<WabtEthereumInterface>(interface);
```
Comment (axic) [inline src/wabt.cpp]: This made sure it was free'd.

File: src/wabt.cpp (hunk 4 above)
Comment (chfast) [reply src/wabt.cpp]: Yes, reported by ASan. I will fix it.

**Comment by axic** [comment]
This killed macos support:
```
[ 13%] Performing configure step for 'wabt'
-- Using prebuilt re2c lexer
CMake Error at /usr/local/Cellar/cmake/3.12.4/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find PythonInterp: Found unsuitable version "1.4", but required
  is at least "2.7" (found /usr/local/bin/python2.7)
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.12.4/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:376 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/Cellar/cmake/3.12.4/share/cmake/Modules/FindPythonInterp.cmake:159 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  CMakeLists.txt:433 (find_package)
```

**Pull request merged**